interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp

```diff
@@ -124,8 +124,8 @@ void IncrementalExecutor::shuttingDown() {
   }
 }
 
-void IncrementalExecutor::AddAtExitFunc(void (*func) (void*), void* arg,
-                                        llvm::Module* M) {
+void IncrementalExecutor::AddAtExitFunc(void (*func)(void*), void* arg,
+                                        std::shared_ptr<llvm::Module> M) {
```

Comment:
Should this entry in the exit functions *own* the module? If not `const weak_ptr&`

Reply:
Make it a `const shared_ptr&`; I agree it's a good sanity check that the `Module` is not owned by the AtExitFuncs anymore.

Reply:
I don't find this pattern in the codebase of llvm. I assume nowadays compilers can use move semantics.

Reply:
Does AddAtExitFunc has any 'possibility' of recording the pointer or is it just using it during its execution?  If it is the later, you might consider just passing the raw pointer (or const shared_ptr&).

Reply:
Vassil wants the ExitFunc to keep ref counts. See https://stackoverflow.com/questions/37610494/passing-const-shared-ptrt-versus-just-shared-ptrt-as-parameter why that should be fixed in the codebase of llvm :-)

Reply:
Done.